examples/DataApi-NFT/template/modules/shared_data/src/main/scala/com/my/nft_example/shared_data/Data.scala

```diff
@@ -0,0 +1,142 @@
+package com.my.nft_example.shared_data
+
+import cats.data.NonEmptyList
+import cats.effect.IO
+import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+import io.circe.{Decoder, Encoder}
+import org.tessellation.currency.dataApplication.dataApplication.DataApplicationValidationErrorOr
+import org.tessellation.currency.dataApplication.{DataApplicationValidationError, DataState, DataUpdate, L1NodeContext}
+import org.tessellation.security.signature.Signed
+import cats.syntax.all._
+import com.my.nft_example.shared_data.Combiners.{combineMintCollection, combineMintNFT, combineTransferCollection, combineTransferNFT}
+import com.my.nft_example.shared_data.Errors.{CouldNotGetLatestCurrencySnapshot, CouldNotGetLatestState}
+import com.my.nft_example.shared_data.Utils.{customStateDeserialization, customStateSerialization, customUpdateDeserialization, customUpdateSerialization}
+import com.my.nft_example.shared_data.Validations.{mintCollectionValidations, mintNFTValidations, mintNFTValidationsWithSignature, transferCollectionValidations, transferCollectionValidationsWithSignature, transferNFTValidations, transferNFTValidationsWithSignature}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import org.tessellation.schema.address.Address
+import org.tessellation.security.SecurityProvider
+
+object Data {
+  @derive(decoder, encoder)
+  sealed trait NFTUpdate extends DataUpdate
+
+  @derive(decoder, encoder)
+  case class MintCollection(name: String) extends NFTUpdate
+
+  @derive(decoder, encoder)
+  case class MintNFT(owner: Address, collectionId: String, nftId: Long, uri: String, name: String, description: String, metadata: Map[String, String]) extends NFTUpdate
+
+  @derive(decoder, encoder)
+  case class TransferCollection(fromAddress: Address, toAddress: Address, collectionId: String) extends NFTUpdate
+
+  @derive(decoder, encoder)
+  case class TransferNFT(fromAddress: Address, toAddress: Address, collectionId: String, nftId: Long) extends NFTUpdate
+
+  @derive(decoder, encoder)
+  sealed trait CollectionState {
```

Comment:
So I haven't gone through whole PR yet and may have missed it -- but I strongly **strongly** recommend implementing validators on the size of all these fields.

String or Maps can be used to dump huge amount of data in here. I'm not sure if already implemented also, but even the endpoints should have a default config to reject requests of more than a certain byte size (which can help prevent even attempting to decode them.)

Every single field that is a primitive type that can encode data (i.e. something bigger than long,) needs a size validator to make sure users aren't dumping enormous amounts of data in them.